File: android/src/main/kotlin/com/prongbang/lazysecret/LazySecretMethodChannel.kt
package com.prongbang.lazysecret

import com.goterl.lazysodium.interfaces.SecretBox
import com.goterl.lazysodium.utils.KeyPair
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel

class LazySecretMethodChannel(
    private val lazySecret: LazySecret
) {

    companion object {
        private const val PARAMETER_NOT_FOUND_ERROR_CODE = "NF001"
        private const val TO_HEX_METHOD = "toHex"
        private const val TO_BIN_METHOD = "toBin"
        private const val BOX_BEFORE_NM_METHOD = "cryptoBoxBeforeNm"
        private const val SECRET_BOX_EASY_METHOD = "cryptoSecretBoxEasy"
        private const val SECRET_BOX_OPEN_EASY_METHOD = "cryptoSecretBoxOpenEasy"
        private const val CREATE_KEY_PAIR_METHOD = "cryptoKxKeyPair"
        private const val SECRET_BOX_KEY_BYTES_METHOD = "cryptoSecretBoxKeyBytes"
        private const val SECRET_BOX_NONCE_BYTES_METHOD = "cryptoSecretBoxNonceBytes"
        private const val SECRET_BOX_MAC_BYTES_METHOD = "cryptoSecretBoxMacBytes"
        private const val RANDOM_BYTES_BUF_METHOD = "randomBytesBuf"
    }

    fun methodCall(call: MethodCall, result: MethodChannel.Result) {
        if (call.method == SECRET_BOX_KEY_BYTES_METHOD) {
            result.success(SecretBox.KEYBYTES)
        } else if (call.method == SECRET_BOX_NONCE_BYTES_METHOD) {
            result.success(SecretBox.NONCEBYTES)
        } else if (call.method == SECRET_BOX_MAC_BYTES_METHOD) {
            result.success(SecretBox.MACBYTES)
        } else if (call.method == TO_HEX_METHOD) {
            val hasBin = call.hasArgument("bin")
            if (hasBin) {
                val bin = call.argument<ByteArray>("bin") ?: byteArrayOf()
                val hex = lazySecret.toHex(bin).lowercase()
                result.success(hex)
            } else {
                val exception = Exception("Byte arrays not found")
                result.error(PARAMETER_NOT_FOUND_ERROR_CODE, exception.message, exception)
            }
        } else if (call.method == TO_BIN_METHOD) {
            val hasHex = call.hasArgument("hex")
            if (hasHex) {
                val hex = call.argument<String>("hex") ?: ""
                result.success(lazySecret.toBin(hex))
            } else {
                val exception = Exception("Hex string not found")
                result.error(PARAMETER_NOT_FOUND_ERROR_CODE, exception.message, exception)
            }
        } else if (call.method == RANDOM_BYTES_BUF_METHOD) {
            val hasSize = call.hasArgument("size")
            if (hasSize) {
                val size = call.argument<Int>("size") ?: 0
                val bytes = lazySecret.randomBytesBuf(size)
                result.success(bytes)
            } else {
                val exception = Exception("Size not found")
                result.error(PARAMETER_NOT_FOUND_ERROR_CODE, exception.message, exception)
            }
        } else if (call.method == CREATE_KEY_PAIR_METHOD) {
            val keyPair = lazySecret.cryptoKxKeypair()
            result.success(
                mapOf(
                    "pk" to keyPair.publicKey.asHexString,
                    "sk" to keyPair.secretKey.asHexString,
                )
            )
        } else if (call.method == BOX_BEFORE_NM_METHOD) {
            val hasPk = call.hasArgument("pk")
            val hasSk = call.hasArgument("sk")
            if (hasPk && hasSk) {
                val pk = call.argument<String>("pk") ?: ""
                val sk = call.argument<String>("sk") ?: ""
                val keyPair = KeyPair(
                    lazySecret.fromHexString(pk),
                    lazySecret.fromHexString(sk),
                )
                val sharedKey = lazySecret.cryptoBoxBeforeNm(keyPair).lowercase()
                result.success(sharedKey)
            } else {
                val exception = Exception("Pk or Sk string not found")
                result.error(PARAMETER_NOT_FOUND_ERROR_CODE, exception.message, exception)
            }
        } else if (call.method == SECRET_BOX_EASY_METHOD) {
            val hasText = call.hasArgument("plaintext")
            val hasNonce = call.hasArgument("nonce")
            val hasKey = call.hasArgument("key")
            if (hasText && hasNonce && hasKey) {
                val plaintext = call.argument<String>("plaintext") ?: ""
                val nonce = call.argument<String>("nonce") ?: ""
                val key = call.argument<String>("key") ?: ""
                val data = lazySecret.cryptoSecretBoxEasy(
                    plaintext = plaintext,
                    nonce = lazySecret.toBin(nonce),
                    key = lazySecret.fromHexString(key),
                ).lowercase()
                result.success(data)
            } else {
                val exception = Exception("PlainText or nonce or key string not found")
                result.error(PARAMETER_NOT_FOUND_ERROR_CODE, exception.message, exception)
            }
        } else if (call.method == SECRET_BOX_OPEN_EASY_METHOD) {
            val hasText = call.hasArgument("ciphertext")
            val hasNonce = call.hasArgument("nonce")
            val hasKey = call.hasArgument("key")
            if (hasText && hasNonce && hasKey) {
                val text = call.argument<String>("ciphertext") ?: ""
                val nonce = call.argument<String>("nonce") ?: ""
                val key = call.argument<String>("key") ?: ""
                val data = lazySecret.cryptoSecretBoxOpenEasy(
                    ciphertext = text,
                    nonce = lazySecret.toBin(nonce),
                    key = lazySecret.fromHexString(key),
                )
                result.success(data)
            } else {
                val exception = Exception("CipherText or nonce or key string not found")
                result.error(PARAMETER_NOT_FOUND_ERROR_CODE, exception.message, exception)
            }
        } else {
            result.notImplemented()
        }
    }
}

File: ios/Classes/LazySecretMethodChannel.swift
//
//  LazySecretMethodChannel.swift
//  lazysecret
//
//  Created by prongbang on 13/9/2566 BE.
//

import Flutter
import Sodium

public class LazySecretMethodChannel {
    
    private let lazysecret: LazySecret
    
    public init(lazysecret: LazySecret) {
        self.lazysecret = lazysecret
    }
    
    enum ErrorCode {
        static let parameterNotFoundErrorCode = "NF001"
    }
    
    enum Method {
        static let toHexMethod = "toHex"
        static let toBinMethod = "toBin"
        static let boxBeforeNmMethod = "cryptoBoxBeforeNm"
        static let secretBoxEasyMethod = "cryptoSecretBoxEasy"
        static let secretBoxOpenEasyMethod = "cryptoSecretBoxOpenEasy"
        static let createKeyPairMethod = "cryptoKxKeyPair"
        static let secretBoxKeyBytesMethod = "cryptoSecretBoxKeyBytes"
        static let secretBoxNonceBytesMethod = "cryptoSecretBoxNonceBytes"
        static let secretBoxMacBytesMethod = "cryptoSecretBoxMacBytes"
        static let randomBytesBufMethod = "randomBytesBuf"
    }
    
    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case Method.toHexMethod:
            let message = "Byte arrays not found"
            guard let args = call.arguments as? Dictionary<String, Any> else {
                result(messageError(message: message))
                return
            }
            guard let binAny = args["bin"] else {
                result(messageError(message: message))
                return
            }
            guard let bytes = parseBytes(from: binAny) else {
                result(messageError(message: message))
                return
            }
            let data = lazysecret.toHex(bytes: bytes)
            result(data)
            break
        case Method.toBinMethod:
            let message = "Hex string not found"
            guard let args = call.arguments as? Dictionary<String, Any> else {
                result(messageError(message: message))
                return
            }
            guard let hexAny = args["hex"] else {
                result(messageError(message: message))
                return
            }
            guard let hex = parseString(from: hexAny) else {
                result(messageError(message: message))
                return
            }
            let data = lazysecret.toBin(hexString: hex)
            result(data)
            break
        case Method.boxBeforeNmMethod:
            let message = "Pk or Sk string not found"
            guard let args = call.arguments as? Dictionary<String, Any> else {
                result(messageError(message: message))
                return
            }
            guard let pkAny = args["pk"] else {
                result(messageError(message: message))
                return
            }
            guard let skAny = args["sk"] else {
                result(messageError(message: message))
                return
            }
            guard let pk = parseString(from: pkAny) else {
                result(messageError(message: message))
                return
            }
            guard let sk = parseString(from: skAny) else {
                result(messageError(message: message))
                return
            }
            let keyPair = KeyPair(
                pk: lazysecret.toBin(hexString: pk),
                sk: lazysecret.toBin(hexString: sk)
            )
            let data = lazysecret.cryptoBoxBeforeNm(keyPair: keyPair)
            result(data)
            break
        case Method.secretBoxEasyMethod:
            let message = "PlainText or nonce or key string not found"
            guard let args = call.arguments as? Dictionary<String, Any> else {
                result(messageError(message: message))
                return
            }
            guard let plaintextAny = args["plaintext"] else {
                result(messageError(message: message))
                return
            }
            guard let nonceAny = args["nonce"] else {
                result(messageError(message: message))
                return
            }
            guard let keyAny = args["key"] else {
                result(messageError(message: message))
                return
            }
            guard let plaintext = parseString(from: plaintextAny) else {
                result(messageError(message: message))
                return
            }
            guard let nonce = parseString(from: nonceAny) else {
                result(messageError(message: message))
                return
            }
            guard let key = parseString(from: keyAny) else {
                result(messageError(message: message))
                return
            }
            let data = lazysecret.cryptoSecretBoxEasy(
                plaintext: plaintext,
                nonce: nonce,
                key: key
            )
            result(data)
            break
        case Method.secretBoxOpenEasyMethod:
            let message = "CipherText or nonce or key string not found"
            guard let args = call.arguments as? Dictionary<String, Any> else {
                result(messageError(message: message))
                return
            }
            guard let ciphertextAny = args["ciphertext"] else {
                result(messageError(message: message))
                return
            }
            guard let nonceAny = args["nonce"] else {
                result(messageError(message: message))
                return
            }
            guard let keyAny = args["key"] else {
                result(messageError(message: message))
                return
            }
            guard let ciphertext = parseString(from: ciphertextAny) else {
                result(messageError(message: message))
                return
            }
            guard let nonce = parseString(from: nonceAny) else {
                result(messageError(message: message))
                return
            }
            guard let key = parseString(from: keyAny) else {
                result(messageError(message: message))
                return
            }
            let data = lazysecret.cryptoSecretBoxOpenEasy(
                ciphertext: ciphertext,
                nonce: nonce,
                key: key
            )
            result(data)
            break
        case Method.createKeyPairMethod:
            guard let keyPair = lazysecret.cryptoKxKeyPair() else {
                let data: [String: String] = [:]
                result(data)
                return
            }
            let data: [String: String] = [
                "pk": lazysecret.toHex(bytes: keyPair.pk),
                "sk": lazysecret.toHex(bytes: keyPair.sk),
            ]
            result(data)
            break
        case Method.secretBoxKeyBytesMethod:
            let size = lazysecret.cryptoSecretBoxKeyBytes()
            result(size)
            break
        case Method.secretBoxNonceBytesMethod:
            let size = lazysecret.cryptoSecretBoxNonceBytes()
            result(size)
            break
        case Method.secretBoxMacBytesMethod:
            let size = lazysecret.cryptoSecretBoxMacBytes()
            result(size)
            break
        case Method.randomBytesBufMethod:
            let message = "Size not found"
            guard let args = call.arguments as? Dictionary<String, Any> else {
                result(messageError(message: message))
                return
            }
            guard let sizeAny = args["size"] else {
                result(messageError(message: message))
                return
            }
            guard let size = parseInt(from: sizeAny) else {
                result(messageError(message: message))
                return
            }
            let data = lazysecret.randomBytesBuf(size: size)
            result(data)
            break
        default:
            result(FlutterMethodNotImplemented)
        }
    }
    
    func messageError(message: String) -> FlutterError {
        return FlutterError(
            code: ErrorCode.parameterNotFoundErrorCode,
            message: message,
            details: nil
        )
    }
    
    func parseInt(from value: Any) -> Int? {
        if let intValue = value as? Int {
            return intValue
        } else if let stringValue = value as? String, let intValue = Int(stringValue) {
            return intValue
        } else {
            return nil
        }
    }
    
    func parseBytes(from value: Any) -> Bytes? {
        if let byteArray = value as? Bytes {
            return byteArray
        } else if let intValue = value as? Int {
            return [UInt8(intValue)]
        } else if let stringValue = value as? String, let intValue = Int(stringValue) {
            return [UInt8(intValue)]
        } else {
            return nil
        }
    }
    
    func parseString(from value: Any) -> String? {
        if let stringValue = value as? String {
            return stringValue
        } else if let intValue = value as? Int {
            return String(intValue)
        } else if let doubleValue = value as? Double {
            return String(doubleValue)
        } else {
            return nil
        }
    }
    
}
